core/src/main/java/com/linecorp/armeria/internal/common/util/ObjectCollectingUtil.java

```diff
@@ -82,15 +85,20 @@
      * Collects objects published from the specified {@link Publisher}.
      *
      * @param publisher publishes objects
+     * @param ctx {@link ServiceRequestContext}
      * @return a {@link CompletableFuture} which will complete when all published objects are collected
      */
-    public static CompletableFuture<Object> collectFrom(Publisher<?> publisher) {
+    @SuppressWarnings("unchecked")
+    public static CompletableFuture<Object> collectFrom(Publisher<?> publisher, ServiceRequestContext ctx) {
         requireNonNull(publisher, "publisher");
+        if (publisher instanceof StreamMessage) {
+            return (CompletableFuture<Object>) (CompletableFuture<?>) ((StreamMessage<?>) publisher).collect();
```

Comment:
It would be better to abort `StreamMessage` when cancelling.
```java
ctx.whenRequestCancelling().thenAccept(cause -> {
   ((StreamMessage<?>) publisher).abort(cause);
});
```